ports/qemu-riscv/Makefile

```diff
@@ -0,0 +1,120 @@
+include ../../py/mkenv.mk
+-include mpconfigport.mk
+
+# qstr definitions (must come before including py.mk)
+QSTR_DEFS = qstrdefsport.h
+
+# MicroPython feature configurations
+MICROPY_ROM_TEXT_COMPRESSION ?= 1
+
+# include py core make definitions
+include $(TOP)/py/py.mk
+include $(TOP)/extmod/extmod.mk
+
+CROSS_COMPILE ?= riscv64-unknown-elf-
+
+BOARD ?= virt
+LDFLAGS = --gc-sections -Map=$(@:.elf=.map)
+LIBS = $(shell $(CC) $(CFLAGS) -print-libgcc-file-name)
+	
+ifeq ($(BOARD),virt)
+ABI=ilp32
+ARCH=rv32imc_zicsr
+AFLAGS = -mabi=$(ABI) -march=$(ARCH)
+CFLAGS += -mabi=$(ABI) -mlittle-endian -march=$(ARCH) -mcmodel=medany
+CFLAGS += -DQEMU_SOC_VIRT
+LDSCRIPT = virt.ld
+LDFLAGS += -T $(LDSCRIPT) -m elf32lriscv
+SRC_BOARD_O = shared/runtime/gchelper_generic.o setjmp.o
+SRC_BOARD_O += entry.o
+endif
+
+INC += -I.
+INC += -I$(TOP)
+INC += -I$(BUILD)
+
+CFLAGS += $(INC) -Wall -Wpointer-arith -Wdouble-promotion -Wfloat-conversion -Werror -std=gnu99 $(COPT) \
+	 -ffunction-sections -fdata-sections
+CFLAGS += $(CFLAGS_EXTRA)
+
+# Debugging/Optimization
+ifeq ($(DEBUG), 1)
+CFLAGS += -g
+COPT = -O0
+else
+COPT += -Os -DNDEBUG
+endif
+
+SRC_COMMON_C = \
+	interrupts.c \
+	startup.c \
+	uart.c \
+	modmachine.c \
+	shared/libc/string0.c \
+	shared/runtime/sys_stdio_mphal.c \
+
+SRC_RUN_C = \
+	main.c \
+
+SRC_TEST_C = \
+	test_main.c \
+	lib/tinytest/tinytest.c \
+
+LIB_SRC_C += $(addprefix lib/,\
+	libm/math.c \
+	libm/fmodf.c \
+	libm/nearbyintf.c \
+	libm/ef_sqrt.c \
+	libm/kf_rem_pio2.c \
+	libm/kf_sin.c \
+	libm/kf_cos.c \
+	libm/kf_tan.c \
+	libm/ef_rem_pio2.c \
+	libm/sf_sin.c \
+	libm/sf_cos.c \
+	libm/sf_tan.c \
+	libm/sf_frexp.c \
+	libm/sf_modf.c \
+	libm/sf_ldexp.c \
+	libm/asinfacosf.c \
+	libm/atanf.c \
+	libm/atan2f.c \
+	libm/roundf.c \
+	)
+
+OBJ_COMMON =
+OBJ_COMMON += $(PY_O)
+OBJ_COMMON += $(addprefix $(BUILD)/, $(SRC_COMMON_C:.c=.o))
+OBJ_COMMON += $(addprefix $(BUILD)/, $(SRC_BOARD_O))
+OBJ_COMMON += $(addprefix $(BUILD)/, $(LIB_SRC_C:.c=.o))
+
+OBJ_RUN =
+OBJ_RUN += $(addprefix $(BUILD)/, $(SRC_RUN_C:.c=.o))
+
+ALL_OBJ_RUN = $(OBJ_COMMON) $(OBJ_RUN)
+
+OBJ_TEST =
+OBJ_TEST += $(addprefix $(BUILD)/, $(SRC_TEST_C:.c=.o))
+
+ALL_OBJ_TEST = $(OBJ_COMMON) $(OBJ_TEST)
+
+# All object files, needed to get dependencies correct
+OBJ = $(OBJ_COMMON) $(OBJ_RUN) $(OBJ_TEST)
+
+# List of sources for qstr extraction
+SRC_QSTR += $(SRC_COMMON_C) $(SRC_RUN_C) $(LIB_SRC_C)
+
+all: run
+
+debug: $(BUILD)/firmware.elf
+	qemu-system-riscv32 -machine $(BOARD) -bios none $(QEMU_EXTRA) -nographic -monitor null -semihosting -serial mon:stdio -S -s -kernel $<
+
+run: $(BUILD)/firmware.elf
+	qemu-system-riscv32 -machine $(BOARD) -bios none $(QEMU_EXTRA) -nographic -monitor null -semihosting -kernel $<
+
+## `$(LD)` doesn't seem to like `--specs` for some reason, but we can just use `$(CC)` here.
```

Comment:
This comment doesn't seem to match the lines below, because LD is used.

Sometimes it is actually better to use CC to do the linking stage, because it can pick up the correct libraries.  Not sure if that would be better here or not.

Reply:
I used `qemu-arm`'s makefile as a template so the comment doesn't apply any longer indeed.  If using LD works for both `qemu-arm` and this, I'd rather use that.

Reply:
OK, keep using LD if it works.

Reply:
As per discussion lower down, linking with CC looks like it's going to be the happy path here in the end. I think riscv and multilib are permanently intertwined, and without it you end up with hard-coded paths (at best) and very limited toolchain support (at worst).